Update MySQL JDBC driver to 8.0.29

hashhar opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Update MySQL JDBC driver to 8.0.29

Fixes #15332 Closes #12692

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

Update MySQL JDBC driver to 8.0.29

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# MySQL
* Update MySQL JDBC driver to 8.0.29. This fixes failure when `query` table function contains `WITH` clause. ({issue}`15332`)
```

hashhar commented:
Alternative to https://github.com/trinodb/trino/pull/12692, let's see what tests fail.

hashhar commented:
Same failures as in https://github.com/trinodb/trino/pull/14917, negative date related tests.

I guess we'll need to add validation on Trino side to disallow negative values until the driver is fixed. I'll see how we can report a bug to the driver - I just left a comment at https://github.com/mysql/mysql-connector-j/commit/60553fd2557267f3184b77e1832025f80d79e89a#r88985720 in past.

hashhar commented:
Got it working, added a test to verify driver bug too so that once we update to working driver we can get notified.

I've started the process to file the bug to MySQL - waiting to hear back.

ebyhr commented on plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:
```diff
@@ -456,6 +471,62 @@ public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connect
+        return LongWriteFunction.of(Types.DATE, (statement, index, epochDay) -> {
+            LocalDate localDate = LocalDate.ofEpochDay(epochDay);
+            if (localDate.isBefore(minMySqlDate)) {
+                throw new TrinoException(INVALID_ARGUMENTS, format("Value must be after %s in MySQL: '%s'", minMySqlDate, localDate));
```
Does "must be after %s" mean exclusive? How about "on or after" or "after or equal to"? 
Please ignore this comment if it doesn't make sense.

ebyhr commented on plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:
```diff
@@ -456,6 +471,62 @@ public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connect
+            if (localDate.isBefore(minMySqlDate)) {
+                throw new TrinoException(INVALID_ARGUMENTS, format("Value must be after %s in MySQL: '%s'", minMySqlDate, localDate));
+            }
+            statement.setObject(index, localDate);
```
If we use `PreparedStatement.setString` with bind expression, there's no need to validate in connector side?

hashhar replied:
Good idea, let me try that.

ebyhr commented on plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlLegacyConnectorTest.java:
```diff
@@ -167,4 +167,24 @@ protected void verifyColumnNameLengthFailurePermissible(Throwable e)
+    public void testDateYearOfEraPredicate()
+    {
+        // MySQL 5 returns incorrect results instead of an empty result with drivers newer than 8.0.23
+        // This is overriden since the connector validates the dates now and throws on such case
```
nit: Fix typo overriden → overridden

hashhar commented:
@ebyhr Addressed comments as fixups.

It's using `bindExpression` now - much cleaner - thanks for the suggestion. The error message changed slightly since now the server sees failure in `CAST` instead of inserting to column.

hashhar commented:
This is ready to review. PTAL @ebyhr @electrum
